cdc/sink/manager/kafka/manager.go

```diff
@@ -0,0 +1,154 @@
+type TopicManager struct {
+	client kafka.Client
+	admin  kafka.ClusterAdminClient
+
+	cfg *kafkaconfig.AutoCreateTopicConfig
+
+	topics sync.Map
+
+	lastMetadataRefresh atomic.Int64
+}
+
+// NewTopicManager creates a new topic manager.
+func NewTopicManager(
+	client kafka.Client,
+	admin kafka.ClusterAdminClient,
+	cfg *kafkaconfig.AutoCreateTopicConfig,
+) *TopicManager {
+	return &TopicManager{
+		client: client,
+		admin:  admin,
+		cfg:    cfg,
+	}
+}
+
+// Partitions returns the number of partitions of the topic.
+// It may also try to update the topics' information maintained by manager.
+func (m *TopicManager) Partitions(topic string) (int32, error) {
+	err := m.tryRefreshMeta()
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	if partitions, ok := m.topics.Load(topic); ok {
+		return partitions.(int32), nil
+	}
+
+	return m.cfg.PartitionNum, m.CreateTopic(topic)
+}
+
+// tryRefreshMeta try to refresh the topics' information maintained by manager.
+func (m *TopicManager) tryRefreshMeta() error {
+	if time.Since(time.Unix(m.lastMetadataRefresh.Load(), 0)) > time.Minute {
+		topics, err := m.client.Topics()
+		if err != nil {
+			return err
+		}
+
+		for _, topic := range topics {
+			partitions, err := m.client.Partitions(topic)
+			if err != nil {
+				return err
+			}
+			m.tryUpdatePartitionsAndLogging(topic, int32(len(partitions)))
+		}
+		m.lastMetadataRefresh.Store(time.Now().Unix())
+	}
+
+	return nil
+}
+
+// tryUpdatePartitionsAndLogging try to update the partitions of the topic.
+func (m *TopicManager) tryUpdatePartitionsAndLogging(topic string, partitions int32) {
+	oldPartitions, ok := m.topics.Load(topic)
+	if ok {
+		if oldPartitions.(int32) != partitions {
+			m.topics.Store(topic, partitions)
+			log.Info(
+				"update topic partition number",
+				zap.String("topic", topic),
+				zap.Int32("oldPartitionNumber", oldPartitions.(int32)),
+				zap.Int32("newPartitionNumber", partitions),
+			)
+		}
+	} else {
+		m.topics.Store(topic, partitions)
+		log.Info(
+			"store topic partition number",
+			zap.String("topic", topic),
+			zap.Int32("partitionNumber", partitions),
+		)
+	}
+}
+
+// CreateTopic creates a topic with the given name.
+func (m *TopicManager) CreateTopic(topicName string) error {
+	topics, err := m.admin.ListTopics()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Now that we have access to the latest topics' information,
+	// we need to update it here immediately.
+	for topic, detail := range topics {
+		m.tryUpdatePartitionsAndLogging(topic, detail.NumPartitions)
+	}
+	m.lastMetadataRefresh.Store(time.Now().Unix())
+
+	// Maybe our cache has expired information, so we just return it.
+	if _, ok := topics[topicName]; ok {
+		return nil
+	}
+
+	if !m.cfg.AutoCreate {
+		return cerror.ErrKafkaInvalidConfig.GenWithStack(
+			fmt.Sprintf("`auto-create-topic` is false, "+
+				"and %s not found", topicName))
+	}
+
+	err = m.admin.CreateTopic(topicName, &sarama.TopicDetail{
```

Comment:
Could you add some logs to track how long these I/O operations takes? Also for list and delete topic.

Reply:
Done.